proposals/xxxx-work-graphs.md

```diff
@@ -0,0 +1,534 @@
+};
+
+/// @brief Interface for GroupNodeInputRecord and RWGroupNodeInputRecord
+template <typename RecordTy, bool IsRW = false> class GroupNodeInputRecordBase {
+  /// @brief Returns the number of records that have been coalesced into the
+  /// current thread group.
+  ///
+  /// Returns 1..._MaxCount_, where _MaxCount_ is specified by the
+  /// `[MaxRecords(_MaxCount_)]` attribute applied to the parameter declaration.
+  uint Count() const;
+
+  /// @brief Get a copy of the underlying record at the specified index.
+  RecordTy Get(uint Index) const;
+
+  /// @brief Get a writable reference to the underlying record.
+  /// @param Index The record index to access.
+  ///
+  /// Only available for non-RW object variangs. The non-const Get() returns a
+  /// reference to the underlying data.
+  std::enable_if_t<IsRW, RecordTy>::type &Get(uint Index);
+
+  /// @brief Get a copy of the underlying record at the specified index.
+  /// @param Index The record index to access.
+  RecordTy operator[](uint Index) const;
+
+  /// @brief Get a writable reference to the underlying record.
+  /// @param Index The record index to access.
+  ///
+  /// Only available for non-RW object variants. The non-const operator[]
+  /// returns a reference to the underlying data.
+  std::enable_if_t<IsRW, RecordTy>::type &operator[](uint Index);
+};
+
+} // namespace detail
+
+template <typename RecordTy>
+using ThreadNodeInputRecord = detail::NodeInputRecordInterface<RecordTy, false>;
+
+template <typename RecordTy>
+using RWThreadNodeInputRecord =
+    detail::NodeInputRecordInterface<RecordTy, true>;
+
+template <typename RecordTy>
+using DispatchNodeInputRecord =
+    detail::NodeInputRecordInterface<RecordTy, false>;
+
+template <typename RecordTy>
+class RWDispatchNodeInputRecord
+    : public detail::NodeInputRecordInterface<RecordTy, true> {
+  /// @brief Allows thread groups to coordinate reading and writing to a shared
+  /// set of records.
+  /// @returns Returns `false` for thread groups that are not the last to finish
+  /// and `true` for the last thread group to call this method.
+  ///
+  /// This method must be called by all threads in a dispatch or not called at
+  /// all. The call must be in dispatch grid uniform control flow. The callsite
+  /// must be uniform across all threads in all thread groups in the dispatch
+  /// grid. This method may be called at most once per thread.
+  ///
+  /// Any violation of these requirements is undefined behavior.
+  ///
+  /// This method returns `false` for thread groups that are not the last to
+  /// finish and these thread groups are not allowed to read or write to the
+  /// input. Reading or writing the input after this call on a thread that
+  /// returned `false` is undefined behavior.
+  ///
+  /// This method returns `true` for the last thread group to finish. That
+  /// thread group can continue reading and writing to the input.
+  bool FinishedCrossGroupSharing();
```

Comment:
That's a lot of text in comments in a pseudo-code block. You'd have more formatting options if you put it outside in metadata. The backticks aren't going to do anything as-is.

Reply:
The intent for this was that the comments are doxygen formatted (hence the
@<tag> tags). Doxygen does support markdown formatting. These header burbs are
actually complete enough that we could probably generate a doxygen header doc
from them.